README.md
# cryptedForage

[![Build Status](https://travis-ci.org/cryptedForage/cryptedForage.svg?branch=master)](http://travis-ci.org/cryptedForage/cryptedForage)
[![NPM version](https://badge.fury.io/js/cryptedforage.svg)](http://badge.fury.io/js/cryptedforage)
[![Dependency Status](https://img.shields.io/david/cryptedForage/cryptedForage.svg)](https://david-dm.org/cryptedForage/cryptedForage)
[![npm](https://img.shields.io/npm/dm/cryptedforage.svg?maxAge=2592000)](https://npmcharts.com/compare/cryptedforage?minimal=true)
[![jsDelivr Hits](https://data.jsdelivr.com/v1/package/npm/cryptedforage/badge?style=rounded)](https://www.jsdelivr.com/package/npm/cryptedforage)

cryptedForage is a fast and simple storage library for JavaScript. cryptedForage
improves the offline experience of your web app by using asynchronous storage
(IndexedDB or WebSQL) with a simple, `localStorage`-like API.

cryptedForage uses localStorage in browsers with no IndexedDB or
WebSQL support. See [the wiki for detailed compatibility info][supported browsers].

To use cryptedForage, just drop a single JavaScript file into your page:

```html
<script src="cryptedforage/dist/cryptedforage.js"></script>
<script>cryptedforage.getItem('something', myCallback);</script>
```

Try the [live example](http://codepen.io/thgreasi/pen/ojYKeE).

Download the [latest cryptedForage from GitHub](https://github.com/cryptedForage/cryptedForage/releases/latest), or install with
[npm](https://www.npmjs.com/):

```bash
npm install cryptedforage
```

cryptedForage is compatible with [browserify](http://browserify.org/).

[supported browsers]: https://github.com/cryptedForage/cryptedForage/wiki/Supported-Browsers-Platforms

## Support

Lost? Need help? Try the
[cryptedForage API documentation](https://cryptedforage.github.io/cryptedForage). [cryptedForage API 文档也有中文版。](https://cryptedforage.docschina.org)

If you're having trouble using the library, running the tests, or want to contribute
to cryptedForage, you can visit
[irc.freenode.net](https://freenode.net/) and head to the `#cryptedforage`
channel to ask questions about cryptedForage.

You can ask [**tofumatt**][tofumatt] about cryptedForage in IRC, who
is usually online from 10am-8pm GMT.

[tofumatt]: http://tofumatt.com/

## Safari 10.1+

Since Safari 10.1 we default to IndexedDB; see the [CHANGELOG](https://github.com/cryptedForage/cryptedForage/blob/master/CHANGELOG.md) for more info.

# How to use cryptedForage

## Callbacks vs Promises

Because cryptedForage uses async storage, it has an async API.
It's otherwise exactly the same as the
[localStorage API](https://hacks.mozilla.org/2009/06/localstorage/).

cryptedForage has a dual API that allows you to either use Node-style callbacks
or [Promises](https://www.promisejs.org/). If you are unsure which one is right for you, it's recommended to use Promises.

Here's an example of the Node-style callback form:

```js
cryptedforage.setItem('key', 'value', function(err) {
    // if err is non-null, we got an error
    cryptedforage.getItem('key', function(err, value) {
        // if err is non-null, we got an error. otherwise, value is the value
    });
});
```

And the Promise form:

```js
cryptedforage
    .setItem('key', 'value')
    .then(function() {
        return cryptedforage.getItem('key');
    })
    .then(function(value) {
        // we got our value
    })
    .catch(function(err) {
        // we got an error
    });
```

For more examples, please visit [the API docs](https://cryptedforage.github.io/cryptedForage).

## Storing Blobs, TypedArrays, and other JS objects

You can store any type in cryptedForage; you aren't limited to strings like in
localStorage. Even if localStorage is your storage backend, cryptedForage
automatically does `JSON.parse()` and `JSON.stringify()` when getting/setting
values.

cryptedForage supports storing all native JS objects that can be serialized to
JSON, as well as ArrayBuffers, Blobs, and TypedArrays. Check the
[API docs][api] for a full list of types supported by cryptedForage.

All types are supported in every storage backend, though storage limits in
localStorage make storing many large Blobs impossible.

[api]: https://cryptedforage.github.io/cryptedForage/#data-api-setitem

## Configuration

You can set database information with the `config()` method.
Available options are `driver`, `name`, `storeName`, `version`, `size`, and
`description`.

Example:

```javascript
cryptedforage.config({
    driver: cryptedforage.WEBSQL, // Force WebSQL; same as using setDriver()
    name: 'myApp',
    version: 1.0,
    size: 4980736, // Size of database, in bytes. WebSQL-only for now.
    storeName: 'keyvaluepairs', // Should be alphanumeric, with underscores.
    description: 'some description',
    secret: 'YOUR_SECRET' // will be generated automatically if not specified
});
```

**Note:** you must call `config()` _before_ you interact with your data. This
means calling `config()` before using `getItem()`, `setItem()`, `removeItem()`,
`clear()`, `key()`, `keys()` or `length()`.

## Multiple instances

You can create multiple instances of cryptedForage that point to different stores
using `createInstance`. All the configuration options used by
[`config`](#configuration) are supported.

```javascript
var store = cryptedforage.createInstance({
    name: 'nameHere'
});

var otherStore = cryptedforage.createInstance({
    name: 'otherName'
});

// Setting the key on one of these doesn't affect the other.
store.setItem('key', 'value');
otherStore.setItem('key', 'value2');
```

## RequireJS

You can use cryptedForage with [RequireJS](http://requirejs.org/):

```javascript
define(['cryptedforage'], function(cryptedforage) {
    // As a callback:
    cryptedforage.setItem('mykey', 'myvalue', console.log);

    // With a Promise:
    cryptedforage.setItem('mykey', 'myvalue').then(console.log);
});
```

## Browserify and Webpack

cryptedForage 1.3+ works with both Browserify and Webpack. If you're using an
earlier version of cryptedForage and are having issues with Browserify or
Webpack, please upgrade to 1.3.0 or above.

Webpack will emit a warning about using a prebuilt javascript file which is fine. If you want to remove the warning you should exclude `cryptedforage` from being parsed by webpack using the following conf :

```javascript
module: {
  noParse: /node_modules\/cryptedforage\/dist\/cryptedforage.js/,
  loaders: [...],
```

## TypeScript

If you have the [`allowSyntheticDefaultImports` compiler option](https://www.typescriptlang.org/docs/handbook/compiler-options.html) set to `true` in your [tsconfig.json](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html) (supported in TypeScript v1.8+), you should use:

```javascript
import cryptedForage from 'cryptedforage';
```

Otherwise you should use one of the following:

```javascript
import * as cryptedForage from "cryptedforage";
// or, in case that the typescript version that you are using
// doesn't support ES6 style imports for UMD modules like cryptedForage
import cryptedForage = require("cryptedforage");
```

## Framework Support

If you use a framework listed, there's a cryptedForage storage driver for the
models in your framework so you can store data offline with cryptedForage. We
have drivers for the following frameworks:

*   [AngularJS](https://github.com/ocombe/angular-cryptedForage)
*   [Angular 4 and up](https://github.com/Alorel/ngforage/)
*   [Backbone](https://github.com/cryptedForage/cryptedForage-backbone)
*   [Ember](https://github.com/genkgo/ember-cryptedforage-adapter)
*   [Vue](https://github.com/dmlzj/vlf)

If you have a driver you'd like listed, please
[open an issue](https://github.com/cryptedForage/cryptedForage/issues/new) to have it
added to this list.

## Custom Drivers

You can create your own driver if you want; see the
[`defineDriver`](https://cryptedforage.github.io/cryptedForage/#driver-api-definedriver) API docs.

There is a [list of custom drivers on the wiki][custom drivers].

[custom drivers]: https://github.com/cryptedForage/cryptedForage/wiki/Custom-Drivers

# Working on cryptedForage

You'll need [node/npm](http://nodejs.org/) and
[bower](http://bower.io/#installing-bower).

To work on cryptedForage, you should start by
[forking it](https://github.com/cryptedForage/cryptedForage/fork) and installing its
dependencies. Replace `USERNAME` with your GitHub username and run the
following:

```bash
# Install bower globally if you don't have it:
npm install -g bower

# Replace USERNAME with your GitHub username:
git clone git@github.com:USERNAME/cryptedForage.git
cd cryptedForage
npm install
bower install
```

Omitting the bower dependencies will cause the tests to fail!

## Running Tests

You need PhantomJS installed to run local tests. Run `npm test` (or,
directly: `grunt test`). Your code must also pass the
[linter](http://jshint.com/).

cryptedForage is designed to run in the browser, so the tests explicitly require
a browser environment. Local tests are run on a headless WebKit (using
[PhantomJS](http://phantomjs.org)).

When you submit a pull request, tests will be run against all browsers that
cryptedForage supports on Travis CI using [Sauce Labs](https://saucelabs.com/).

# License

This program is free software; it is distributed under an
[Apache License](https://github.com/cryptedForage/cryptedForage/blob/master/LICENSE).

---

Copyright (c) 2013-2016 [Mozilla](https://mozilla.org)
([Contributors](https://github.com/cryptedForage/cryptedForage/graphs/contributors)).


CHANGELOG.md
# About this CHANGELOG

This file will include all API breakage, new features, and upgrade info in
cryptedForage's lifetime.

### [1.7.3](https://github.com/mozilla/cryptedForage/releases/tag/1.7.3)

* Add `.npmignore` file to reduce package size when installed via npm.

### [1.6](https://github.com/mozilla/cryptedForage/releases/tag/1.6.0)

* Add `dropInstance()` method to cryptedforage.
* Improve IDB driver reliability when a connection gets closed.

### [1.5.7]

* Fix IE8 localStorage support detection.

### [1.5.6]

* Upgrade lie to 3.1.1 to work with yarn.

### [1.5.5]

* Roll back dropInstance breaking change.

### [1.5.4]

* Set `null` as `undefined` (for Edge).

### [1.5.3]

* Check whether localStorage is actually usable.

### [1.5.2]

* Prevent some unnecessary logs when calling `createInstance()`.

### [1.5.1]

* Try to re-establish IDB connection after an InvalidStateError.
* Added Generics to `iterate()` TS Typings.
* Define custom drivers syncronously when `_support` isn't a function.
* Use the custom driver API for internal drivers, which makes possible to override parts of their implementation.

### [1.5](https://github.com/mozilla/cryptedForage/releases/tag/1.5.0)
* **Major storage engine change for Safari**: We now use IndexedDB as the storage engine for Safari v10.1 (and above). This means that **pre-existing Safari** `>= 10.1` users will experience "data loss" **after upgrading your site from a previous version of cryptedForage to v1.5**. In fact no data is lost but the engine will change so cryptedForage will seem empty.
  * You can use the [cryptedForage 1.4 compatibility plugin](https://github.com/cryptedForage/cryptedForage-compatibility-1-4) after the upgrade so that all your Safari users (both old & new) continue to use the WebSQL driver.
  * Alternativelly you can force a connection to WebSQL using [cryptedForage's config](https://cryptedforage.github.io/cryptedForage/#settings-api-setdriver) to either keep using your existing WebSQL database or migrate to IndexedDB.

### [1.4.2](https://github.com/mozilla/cryptedForage/releases/tag/1.4.2)
* Fixes #562.

### [1.4.1](https://github.com/mozilla/cryptedForage/releases/tag/1.4.1)
* Fixes #520; browserify builds work properly

### [1.4](https://github.com/mozilla/cryptedForage/releases/tag/1.4.0)
* Fixes #516; this version should always load the correct driver without that bug.

### [1.3](https://github.com/mozilla/cryptedForage/releases/tag/1.3.0)
* We now use ES6 for our source code and `webpack` to bundle the `dist/` files.

### [1.2](https://github.com/mozilla/cryptedForage/releases/tag/1.2.0)
* Iterate through the entire database using `iterate()`. ([#283](https://github.com/mozilla/cryptedForage/pull/283); fixes [#186](https://github.com/mozilla/cryptedForage/pull/186))

### [1.1](https://github.com/mozilla/cryptedForage/releases/tag/1.1.0)
* Custom drivers can be created using `defineDriver()`. ([#282](https://github.com/mozilla/cryptedForage/pull/282); fixes [#267](https://github.com/mozilla/cryptedForage/pull/267))

### [1.0.3](https://github.com/mozilla/cryptedForage/releases/tag/1.0.3)
* `config()` accepts a new option: `driver`, so users can set the driver during config rather than using `setDriver()`. ([#273](https://github.com/mozilla/cryptedForage/pull/273); fixes [#168](https://github.com/mozilla/cryptedForage/pull/168))

### [1.0](https://github.com/mozilla/cryptedForage/releases/tag/1.0.0)

* It is no longer necessary to queue commands using `ready()` when using RequireJS. ([723cc94e06](https://github.com/mozilla/cryptedForage/commit/723cc94e06af4f5ba4c53fa65524ccd5f6c4432e))
* `setDriver` now accepts an array of drivers to be used, in order of preference, instead of simply a string. The string option is still supported. (eg. now one can use `setDriver(['WebSQL', 'localStorage'])` instead of `setDriver('WebSQL')`)
* node-style, error-first argument method signatures are used for callbacks. Promises don't use error-first method signatures; instead they supply an error to the promise's `reject()` function.

### [0.9](https://github.com/mozilla/cryptedForage/releases/tag/0.9.1)

This release drops support for some legacy browsers, though not actually the
ones you might think. cryptedForage's new policy is to support the current
version of all major browsers plus up to three versions back.

* Add built versions without the Promises polyfill to `dist/` directory. ([#172](https://github.com/mozilla/cryptedForage/pull/172))
* **Drop support for Firefox 3.5. Minimum version is now Firefox 25.** (Technically, Firefox 4+ seems to work.)
* **Drop support for Chrome 31 and below. Minimum version is now Chrome 32.**
* Fix a **lot** of bugs. Especially in Internet Exploder.
* Switch to Mocha tests and test on [Sauce Labs](https://saucelabs.com/).
* Add a `keys()` method. ([#180](https://github.com/mozilla/cryptedForage/pull/180))
* Check for localStorage instead of assuming it's available. ([#183](https://github.com/mozilla/cryptedForage/pull/183))

### [Version 0.8](https://github.com/mozilla/cryptedForage/releases/tag/0.8.1)

* Add support for web workers. ([#144](https://github.com/mozilla/cryptedForage/pull/144), [#147](https://github.com/mozilla/cryptedForage/pull/147)).

### [Version 0.6.1](https://github.com/mozilla/cryptedForage/releases/tag/0.6.1)

* Put built versions back in `dist/` directory.

### [Version 0.6.0](https://github.com/mozilla/cryptedForage/releases/tag/0.6.0)

* Add `cryptedforage.config`. ([#40](https://github.com/mozilla/cryptedForage/pull/140))
* Fix iFrame bug in WebKit. ([#78](https://github.com/mozilla/cryptedForage/issues/78))
* Improve error handling. ([#60](https://github.com/mozilla/cryptedForage/issues/60))
* Remove support for `window.cryptedForageConfig`. ([#135](https://github.com/mozilla/cryptedForage/issues/135))

### [Version 0.4](https://github.com/mozilla/cryptedForage/releases/tag/0.4.0)


src/cryptedforage.js
import idbDriver from './drivers/indexeddb';
import websqlDriver from './drivers/websql';
import localstorageDriver from './drivers/localstorage';
import serializer from './utils/serializer';
import Promise from './utils/promise';
import executeCallback from './utils/executeCallback';
import executeTwoCallbacks from './utils/executeTwoCallbacks';
import includes from './utils/includes';
import isArray from './utils/isArray';
import generateSecret from './utils/generateSecret';

// Drivers are stored here when `defineDriver()` is called.
// They are shared across all instances of cryptedForage.
const DefinedDrivers = {};

const DriverSupport = {};

const DefaultDrivers = {
    INDEXEDDB: idbDriver,
    WEBSQL: websqlDriver,
    LOCALSTORAGE: localstorageDriver
};

const DefaultDriverOrder = [
    DefaultDrivers.INDEXEDDB._driver,
    DefaultDrivers.WEBSQL._driver,
    DefaultDrivers.LOCALSTORAGE._driver
];

const OptionalDriverMethods = ['dropInstance'];

const LibraryMethods = [
    'clear',
    'getItem',
    'iterate',
    'key',
    'keys',
    'length',
    'removeItem',
    'setItem'
].concat(OptionalDriverMethods);

const DefaultConfig = {
    description: '',
    driver: DefaultDriverOrder.slice(),
    name: 'cryptedforage',
    // Default DB size is _JUST UNDER_ 5MB, as it's the highest size
    // we can use without a prompt.
    size: 4980736,
    storeName: 'keyvaluepairs',
    version: 1.0,
    secret: generateSecret(32)
};

function callWhenReady(cryptedForageInstance, libraryMethod) {
    cryptedForageInstance[libraryMethod] = function() {
        const _args = arguments;
        return cryptedForageInstance.ready().then(function() {
            return cryptedForageInstance[libraryMethod].apply(
                cryptedForageInstance,
                _args
            );
        });
    };
}

function extend() {
    for (let i = 1; i < arguments.length; i++) {
        const arg = arguments[i];

        if (arg) {
            for (let key in arg) {
                if (arg.hasOwnProperty(key)) {
                    if (isArray(arg[key])) {
                        arguments[0][key] = arg[key].slice();
                    } else {
                        arguments[0][key] = arg[key];
                    }
                }
            }
        }
    }

    return arguments[0];
}

class CryptedForage {
    constructor(options) {
        for (let driverTypeKey in DefaultDrivers) {
            if (DefaultDrivers.hasOwnProperty(driverTypeKey)) {
                const driver = DefaultDrivers[driverTypeKey];
                const driverName = driver._driver;
                this[driverTypeKey] = driverName;

                if (!DefinedDrivers[driverName]) {
                    // we don't need to wait for the promise,
                    // since the default drivers can be defined
                    // in a blocking manner
                    this.defineDriver(driver);
                }
            }
        }

        this._defaultConfig = extend({}, DefaultConfig);
        this._config = extend({}, this._defaultConfig, options);
        this._driverSet = null;
        this._initDriver = null;
        this._ready = false;
        this._dbInfo = null;

        this._wrapLibraryMethodsWithReady();
        this.setDriver(this._config.driver).catch(() => {});
    }

    // Set any config values for cryptedForage; can be called anytime before
    // the first API call (e.g. `getItem`, `setItem`).
    // We loop through options so we don't overwrite existing config
    // values.
    config(options) {
        // If the options argument is an object, we use it to set values.
        // Otherwise, we return either a specified config value or all
        // config values.
        if (typeof options === 'object') {
            // If cryptedforage is ready and fully initialized, we can't set
            // any new configuration values. Instead, we return an error.
            if (this._ready) {
                return new Error(
                    "Can't call config() after cryptedforage " +
                        'has been used.'
                );
            }

            for (let i in options) {
                if (i === 'storeName') {
                    options[i] = options[i].replace(/\W/g, '_');
                }

                if (i === 'version' && typeof options[i] !== 'number') {
                    return new Error('Database version must be a number.');
                }

                this._config[i] = options[i];
            }

            // after all config options are set and
            // the driver option is used, try setting it
            if ('driver' in options && options.driver) {
                return this.setDriver(this._config.driver);
            }

            return true;
        } else if (typeof options === 'string') {
            return this._config[options];
        } else {
            return this._config;
        }
    }

    // Used to define a custom driver, shared across all instances of
    // cryptedForage.
    defineDriver(driverObject, callback, errorCallback) {
        const promise = new Promise(function(resolve, reject) {
            try {
                const driverName = driverObject._driver;
                const complianceError = new Error(
                    'Custom driver not compliant; see ' +
                        'https://mozilla.github.io/cryptedForage/#definedriver'
                );

                // A driver name should be defined and not overlap with the
                // library-defined, default drivers.
                if (!driverObject._driver) {
                    reject(complianceError);
                    return;
                }

                const driverMethods = LibraryMethods.concat('_initStorage');
                for (let i = 0, len = driverMethods.length; i < len; i++) {
                    const driverMethodName = driverMethods[i];

                    // when the property is there,
                    // it should be a method even when optional
                    const isRequired = !includes(
                        OptionalDriverMethods,
                        driverMethodName
                    );
                    if (
                        (isRequired || driverObject[driverMethodName]) &&
                        typeof driverObject[driverMethodName] !== 'function'
                    ) {
                        reject(complianceError);
                        return;
                    }
                }

                const configureMissingMethods = function() {
                    const methodNotImplementedFactory = function(methodName) {
                        return function() {
                            const error = new Error(
                                `Method ${methodName} is not implemented by the current driver`
                            );
                            const promise = Promise.reject(error);
                            executeCallback(
                                promise,
                                arguments[arguments.length - 1]
                            );
                            return promise;
                        };
                    };

                    for (
                        let i = 0, len = OptionalDriverMethods.length;
                        i < len;
                        i++
                    ) {
                        const optionalDriverMethod = OptionalDriverMethods[i];
                        if (!driverObject[optionalDriverMethod]) {
                            driverObject[
                                optionalDriverMethod
                            ] = methodNotImplementedFactory(
                                optionalDriverMethod
                            );
                        }
                    }
                };

                configureMissingMethods();

                const setDriverSupport = function(support) {
                    if (DefinedDrivers[driverName]) {
                        console.info(
                            `Redefining CryptedForage driver: ${driverName}`
                        );
                    }
                    DefinedDrivers[driverName] = driverObject;
                    DriverSupport[driverName] = support;
                    // don't use a then, so that we can define
                    // drivers that have simple _support methods
                    // in a blocking manner
                    resolve();
                };

                if ('_support' in driverObject) {
                    if (
                        driverObject._support &&
                        typeof driverObject._support === 'function'
                    ) {
                        driverObject._support().then(setDriverSupport, reject);
                    } else {
                        setDriverSupport(!!driverObject._support);
                    }
                } else {
                    setDriverSupport(true);
                }
            } catch (e) {
                reject(e);
            }
        });

        executeTwoCallbacks(promise, callback, errorCallback);
        return promise;
    }

    driver() {
        return this._driver || null;
    }

    getDriver(driverName, callback, errorCallback) {
        const getDriverPromise = DefinedDrivers[driverName]
            ? Promise.resolve(DefinedDrivers[driverName])
            : Promise.reject(new Error('Driver not found.'));

        executeTwoCallbacks(getDriverPromise, callback, errorCallback);
        return getDriverPromise;
    }

    getSerializer(callback) {
        const serializerPromise = Promise.resolve(serializer);
        executeTwoCallbacks(serializerPromise, callback);
        return serializerPromise;
    }

    ready(callback) {
        const self = this;

        const promise = self._driverSet.then(() => {
            if (self._ready === null) {
                self._ready = self._initDriver();
            }

            return self._ready;
        });

        executeTwoCallbacks(promise, callback, callback);
        return promise;
    }

    setDriver(drivers, callback, errorCallback) {
        const self = this;

        if (!isArray(drivers)) {
            drivers = [drivers];
        }

        const supportedDrivers = this._getSupportedDrivers(drivers);

        function setDriverToConfig() {
            self._config.driver = self.driver();
        }

        function extendSelfWithDriver(driver) {
            self._extend(driver);
            setDriverToConfig();

            self._ready = self._initStorage(self._config);
            return self._ready;
        }

        function initDriver(supportedDrivers) {
            return function() {
                let currentDriverIndex = 0;

                function driverPromiseLoop() {
                    while (currentDriverIndex < supportedDrivers.length) {
                        let driverName = supportedDrivers[currentDriverIndex];
                        currentDriverIndex++;

                        self._dbInfo = null;
                        self._ready = null;

                        return self
                            .getDriver(driverName)
                            .then(extendSelfWithDriver)
                            .catch(driverPromiseLoop);
                    }

                    setDriverToConfig();
                    const error = new Error(
                        'No available storage method found.'
                    );
                    self._driverSet = Promise.reject(error);
                    return self._driverSet;
                }

                return driverPromiseLoop();
            };
        }

        // There might be a driver initialization in progress
        // so wait for it to finish in order to avoid a possible
        // race condition to set _dbInfo
        const oldDriverSetDone =
            this._driverSet !== null
                ? this._driverSet.catch(() => Promise.resolve())
                : Promise.resolve();

        this._driverSet = oldDriverSetDone
            .then(() => {
                const driverName = supportedDrivers[0];
                self._dbInfo = null;
                self._ready = null;

                return self.getDriver(driverName).then(driver => {
                    self._driver = driver._driver;
                    setDriverToConfig();
                    self._wrapLibraryMethodsWithReady();
                    self._initDriver = initDriver(supportedDrivers);
                });
            })
            .catch(() => {
                setDriverToConfig();
                const error = new Error('No available storage method found.');
                self._driverSet = Promise.reject(error);
                return self._driverSet;
            });

        executeTwoCallbacks(this._driverSet, callback, errorCallback);
        return this._driverSet;
    }

    supports(driverName) {
        return !!DriverSupport[driverName];
    }

    _extend(libraryMethodsAndProperties) {
        extend(this, libraryMethodsAndProperties);
    }

    _getSupportedDrivers(drivers) {
        const supportedDrivers = [];
        for (let i = 0, len = drivers.length; i < len; i++) {
            const driverName = drivers[i];
            if (this.supports(driverName)) {
                supportedDrivers.push(driverName);
            }
        }
        return supportedDrivers;
    }

    _wrapLibraryMethodsWithReady() {
        // Add a stub for each driver API method that delays the call to the
        // corresponding driver method until cryptedForage is ready. These stubs
        // will be replaced by the driver methods as soon as the driver is
        // loaded, so there is no performance impact.
        for (let i = 0, len = LibraryMethods.length; i < len; i++) {
            callWhenReady(this, LibraryMethods[i]);
        }
    }

    createInstance(options) {
        return new CryptedForage(options);
    }
}

// The actual cryptedForage object that we expose as a module or via a
// global. It's extended by pulling in one of our other libraries.
export default new CryptedForage();
